packages/openapi-parser/src/utils/bundle.ts

```diff
@@ -0,0 +1,219 @@
+
+    return {
+      ok: true,
+      data: JSON.parse(fileContents),
+    }
+  } catch {
+    return {
+      ok: false,
+    }
+  }
+}
+
+/**
+ * Resolves a reference by fetching its content from either a remote URL or local file.
+ * @param ref - The reference string to resolve
+ * @returns A promise that resolves to either the content or an error result
+ * @example
+ * // Resolve a remote URL
+ * await resolveRef('https://example.com/schema.json')
+ * // Resolve a local file
+ * await resolveRef('./schemas/user.json')
+ * // Invalid reference returns { ok: false }
+ * await resolveRef('#/components/schemas/User')
+ */
+async function resolveRef(ref: string): Promise<ResolveResult> {
+  if (isRemoteRef(ref)) {
+    return fetchJson(ref)
+  }
+
+  if (isLocalFileRef(ref)) {
+    return readFile(ref)
+  }
+
+  return {
+    ok: false,
+  }
+}
+
+/**
+ * Retrieves a nested value from an object using an array of property segments.
+ * @param target - The target object to traverse
+ * @param segments - Array of property names representing the path to the desired value
+ * @returns The value at the specified path, or undefined if the path doesn't exist
+ * @example
+ * const obj = { foo: { bar: { baz: 42 } } };
+ * getNestedValue(obj, ['foo', 'bar', 'baz']); // returns 42
+ */
+export function getNestedValue(target: Record<string, any>, segments: string[]) {
+  return segments.reduce<any>((acc, key) => {
+    return acc[key]
+  }, target)
+}
+
+/**
+ * Bundles an OpenAPI specification by resolving all external references.
+ * This function traverses the input object recursively and replaces any external $ref
+ * references with their actual content. External references can be URLs or local files.
+ *
+ * @param input - The OpenAPI specification object to bundle
+ * @returns A promise that resolves when all references have been resolved
+ */
+export function bundle(input: UnknownObject) {
+  // Cache for storing promises of resolved external references (URLs and local files)
+  // to avoid duplicate fetches/reads of the same resource
+  const cache = new Map<string, Promise<ResolveResult>>()
+
+  const bundler = async (root: any) => {
+    if (!root || !isObject(root)) {
+      return
+    }
+
+    await Promise.all(
+      Object.entries(root).map(async ([key, value]) => {
+        if (!isObject(value)) {
+          return
+        }
+
+        if (typeof value === 'object' && '$ref' in value && typeof value['$ref'] === 'string') {
+          const ref = value['$ref']
+
+          if (!isExternalRef(ref)) {
+            return
+          }
+
+          const [prefix, path] = ref.split('/#', 2)
+
+          if (!cache.has(prefix)) {
+            cache.set(prefix, resolveRef(prefix))
+          }
+
+          // Resolve the remote ref
+          const result = await cache.get(prefix)
+
+          if (result.ok) {
+            root[key] = getNestedValue(result.data as any, getSegmentsFromPath(path))
```

Comment:
This does not appear to handle any adjacent fields, correct?  (3.0 and 3.1 Path Item Object field combination, 3.1 Reference Object field overwriting, and 3.1 Schema Object adjacent fields which are complex to handle).

I am guessing it also does not support plain name fragments (declared by `$anchor` or `$dynamicAnchor` in OAS 3.1 Schema Objects.

Note that `$dynamicAnchor` anchors are usable as normal fragments with `$ref`, they just _also_ have special `$dynamicRef` behavior.  I am not suggesting that you can handle `$dynamicRef` here (it's theoretically possible, but very complex and can increase the size of your document exponentially).

I'm assuming no support for `$id` either?